File: src/java/magmac/app/compile/rule/OrRule.java
package magmac.app.compile.rule;

import magmac.api.collect.list.List;
import magmac.api.iter.Iter;
import magmac.app.compile.error.CompileResult;
import magmac.app.compile.error.context.Context;
import magmac.app.compile.error.context.NodeContext;
import magmac.app.compile.error.context.StringContext;
import magmac.app.compile.node.Node;

import java.util.function.Function;

public record OrRule(List<Rule> rules) implements Rule {
    private static <T> OrState<T> foldElement(OrState<T> state, Rule rule, Function<Rule, CompileResult<T>> mapper) {
        return mapper.apply(rule).match(state::withValue, state::withError);
    }

    private <T> CompileResult<T> foldAll(Function<Rule, CompileResult<T>> mapper, Context context) {
        Iter<Rule> ruleIter = this.rules.iter();
        OrState<T> initial = new OrState<T>();
        return ruleIter.fold(initial, (OrState<T> state, Rule rule) -> OrRule.foldElement(state, rule, mapper))
                .toResult(context);
    }

    @Override
    public CompileResult<Node> lex(String input) {
        return this.foldAll((Rule rule1) -> rule1.lex(input), new StringContext(input));
    }

    @Override
    public CompileResult<String> generate(Node node) {
        return this.foldAll((Rule rule1) -> rule1.generate(node), new NodeContext(node));
    }
}

File: src/web/magmac/app/compile/rule/OrRule.ts
import { List } from "../../../../magmac/api/collect/list/List";
import { Iter } from "../../../../magmac/api/iter/Iter";
import { CompileResult } from "../../../../magmac/app/compile/error/CompileResult";
import { Context } from "../../../../magmac/app/compile/error/context/Context";
import { NodeContext } from "../../../../magmac/app/compile/error/context/NodeContext";
import { StringContext } from "../../../../magmac/app/compile/error/context/StringContext";
import { Node } from "../../../../magmac/app/compile/node/Node";
import { Function } from "../../../../java/util/function/Function";
export class OrRule {
	foldElement(state : OrState<T>, rule : Rule, mapper : Function<Rule, CompileResult<T>>) : OrState<T> {return mapper.apply( rule).match( state.withValue, state.withError);;}
	foldAll(mapper : Function<Rule, CompileResult<T>>, context : Context) : CompileResult<T> {ruleIter : Iter<Rule>=this.rules.iter( );initial : OrState<T>=new OrState<T>( );return ruleIter.fold( initial, 0).toResult( context);;}
	lex(input : String) : CompileResult<Node> {return this.foldAll( 0, new StringContext( input));;}
	generate(node : Node) : CompileResult<String> {return this.foldAll( 0, new NodeContext( node));;}
}
